### PR TITLE
Adds configuration file adapter

### PR DESCRIPTION
**Pull request by humanumbrella**
This will allow the transforming of Unchained and Caravan multisig config files into Coldcard multisig config files. Some unit tests to make sure config files have necessary pieces are included.

### CODE REVIEW

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:91]: Should these field validations be in the constructor instead? If certain fields are missing at initialization then surely it's an invalid config which should throw before anything with it is used. This also avoids having to do any validation with any other methods we might have and we can assume that an initialized `ColdcardMultisigConfig` class is good to use.

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:76]: Why leave these in the `jsonConfig`? to me it feels like once initialized from the json that's passed in, these become properties of the config itself. `configInstance.jsonconfig.extendedPublicKeys` feels a little unwieldy if you're trying to access. This becomes even nicer for internal access, e.g. `this.extendedPublicKeys`. In terms of how it reads, it feels more natural too (i.e. "these are the xpubs of this coldcard config class")

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:73]: Something that feels like it would be nice to have are methods that are named like `fromColdcardConfig` and `toColdcardConfig`. The `from` would be a tool to parse what is retrieved from the config file and turn it into a class that has some actual functionality. The `to` method is basically what I'm gathering `adapt` does. 

But maybe this is more of an argument for what's been discussed elsewhere about having a general purpose `MultisigWalletConfig` class which in addition to the above methods could be a lot more useful and flexible and include things like `toElectrumConfig` (and `from`), `toCaravanConfig`, etc. Doing it this way has a nice added benefit of once you've imported/converted, uc-wallets/bitcoin is then providing a uniform interface for interacting with a multisig wallet config that can be passed around and converted as necessary. 

Starting with what's here as a base and just generalizing the naming, even if it only works for coldcard for now shouldn't involve too much while offering more flexibility for the long term.

File: src/coldcard.test.js
Comment (bucko13) [inline src/coldcard.test.js:35]: Any reason why these JSON objects need to be duplicated almost exactly? Since there's already some parsing/stringifying and copying below, couldn't there just be one as a base that is adjusted as necessary for each specific test?

File: src/coldcard.js
Comment (humanumbrella) [reply src/coldcard.js:73]: This sounds good on the surface but I don't how it would work in practice given the missing information (bip32 paths), for example, in the current Coldcard config. The `fromColdcardConfig` e.g. does not give enough information for us to create a CaravanConfig or an ElectrumConfig.

Comment (bucko13) [reply src/coldcard.js:73]: But then this can be handled in the conversion. If you import a config via `fromColdcardConfig`, we could then have methods to add missing information, which the consumer of this library could decide any number of ways they want to retrieve it. Then if a `.toElectrumConfig` is called if it's not available on the current object, you would get an error (and a very specific error telling you why an electrum config isn't possible to generate.

Comment (bucko13) [reply src/coldcard.js:73]: I actually had to do something similar on another project when dealing with LSATs (lightning service authentication tokens). Basically there were a lot of different ways to consume and generate lsats. They are typically, like the Coldcard Config from what I can tell, sent in http headers as text encoded in a particular way. But I needed a way so that specific information could be consumed and easily read. So, this tool would read the plain text, learn where things were broken up and then add as properties on the object. If you wanted to convert back as an LSAT or as a challenge, the conversion method determines if it has enough information and throws a descriptive error if it doesn't. 

Here's the main file I'm thinking about: https://github.com/Tierion/lsat-js/blob/master/src/lsat.ts
You can see in [fromChallenge](https://github.com/Tierion/lsat-js/blob/master/src/lsat.ts#L347) it will look for all the text elements it expects.
Meanwhile [toChallenge](https://github.com/Tierion/lsat-js/blob/master/src/lsat.ts#L245) throws if it doesn't have an invoice property attached to it.

Anyway, sorry to bring a totally random and unrelated project into this. It's just something I was working on relatively recently and felt relevant. In short, I think the ability for this Class to actually have the ability to tell us it can't be converted to an ElectrumConfig is super powerful as a tool and feels like it belongs together with the other conversion classes since the config, unlike the keystore classes, are kind of supposed to be interoperable in my opinion.

Comment (humanumbrella) [reply src/coldcard.js:73]: Yeah - I hear that. I'm open to more discussion 1on1 or with a wider audience, but I'd argue that if you want an Electrum config from a Coldcard, there's a path on the device for that (Multisig Wallet Details > Electrum Wallet) and it'll output the file you want to the SD card.

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:88]: Is this error technically correct? If it's JSON it would be text right? and then it would have to be parsed first in order to access the properties. In either case, it might be a good idea to have some type checks, support both, and then attempt to `JSON.parse` if it's a string, catching an error to say it's mis-configured.

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:81]: How about checking if it's a hex rather than that it is not 'Unknown'. If there's a length restriction, that could be added as well.

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:76]: This is partly just a personal taste thing, but I think it's more readable to do these conditionals one at a type (unless they are dependent on each other).

i.e. 
```
if (jsonConfig.name && jsonConfig.uuid)
  this.name = jsonConfig.uuid || jsonConfig.name
```
An added benefit of this is that if there are any specific properties that are absolutely required vs some that aren't, you can throw in that conditional and say what's missing.

Comment (bucko13) [reply src/coldcard.js:76]: Also for this section, do you have any feelings on also including type checking? Even though JS obviously isn't strictly typed, you can add asserts that will throw to make sure everything is kosher. This also makes this initialization a very effective validator pretty much out of the gate (which you want anyway since the class is useless if any field is not formatted correctly).

File: src/coldcard.test.js
Comment (bucko13) [inline src/coldcard.test.js:4]: Two other suggestions:
1. I'm not sure the quotations around the key names are necessary since this is still technically an object and not a JSON string, so it could just make this easier to maintain.
2. It might be a good idea to put this in a `beforeEach` block so that the value is reset after each test to avoid tests polluting each other.

File: src/coldcard.js
Comment (bucko13) [reply src/coldcard.js:73]: What about for someone not using a coldcard though? What if I want to go from Electrum to Caravan? It also could be useful in other UI settings.

Comment (bucko13) [reply src/coldcard.js:73]: What if you have a UI like in caravan that lets you compose a config that you want to be able to import into either Coldcard or Electrum? You could have the form just pass the object to this class and depending on what is selected, spit out the correctly formatted text/file for Caravan, Coldcard, Electrum, or any other format that ends up getting created which we could extend this base class to support.

File: src/coldcard.test.js
Comment (humanumbrella) [reply src/coldcard.test.js:4]: As for 1 - noted - I just pasted in what comes from unchained external spend JSON. updated
2 - definitely.

File: src/coldcard.test.js
Comment (bucko13) [inline src/coldcard.test.js:15]: this might've been causing the issues with the spread operator since it's not a deep copy and these are objects within objects.

**Comment by waldenraines** [comment]
@humanumbrella @bucko13 what is the status of this?

File: src/coldcard.test.js
Comment (bucko13) [inline src/coldcard.test.js:35]: This doesn't need to be done in this PR necessarily, but I wonder if these would be useful in a fixtures file/directory? Might be something that's used later in other tests. Wouldn't even be surprised if there were some in the coldcard repo, which would be good to standardize on.

Not a blocking comment, just something that occurred to me looking at this again.

File: src/coldcard.test.js
Comment (bucko13) [inline src/coldcard.test.js:61]: This can just be initialized empty. `let jsonConfigCopy;` looks a little cleaner and is pretty standard in my experience. It'll get assigned in the beforeEach.

**Review by bucko13 — COMMENTED** [review]
Just some small nits, really just optional comments. I still think it's worth discussing the Config class further, but otherwise this LGTM (and if we decide to keep it in the current format, then this also LGTM).

File: src/coldcard.test.js
Comment (humanumbrella) [reply src/coldcard.test.js:35]: Yes I think updating the fixtures in `unchained-bitcoin` with PSBT and signed PSBT will be necessary and improve confidence with known transactions, potentially even add it to the test suite.

File: src/coldcard.test.js
Comment (humanumbrella) [reply src/coldcard.test.js]: Got it

**Review by waldenraines — APPROVED** [review]
LGTM but will wait on @bucko13's approval before merging.

File: src/coldcard.test.js
Comment (bucko13) [reply src/coldcard.test.js:35]: Would you mind adding this as an issue ticket to unchained-bitcoin and mentioning what specifically needs to be moved over? It might be best once this is merged in so there's something to link to but just wanted to mention here so we don't forget about it.

Comment (bucko13) [reply src/coldcard.test.js:35]: I'd even support starting by adding a fixtures file here that we can eventually migrate to unchained-bitcoin when psbt support gets better built out there. Again, it shouldn't be a blocker necessarily, but I could see this being perpetually de-prioritized and more likely to have more constants that should be in fixtures added in as we continue building out.

File: src/coldcard.js
Comment (bucko13) [inline src/coldcard.js:77]: Marked this in the other PR (sorry!), but nit around the `else`

**Comment by bucko13** [comment]
I think there are some overlapping comments now from the other PR #41 (mainly around fixtures, some formatting, and jsdoc comments) but in addition I do think the comments around the configuration adapter still warrants a discussion. I feel pretty strongly about this being more general functionality and think it would provide more long term utility creating such a tool. If there's a deeper discussion around this and we still come out on the side of keeping this a ColdCard specific tool, then let's move forward that way, but I still think it warrants a discussion where we're making a fully informed decision one way or another because this will more likely than not be the decision we're stuck with in the future (i.e. if we change our mind in the future, it's unlikely we'd spend the time unwinding). In the hopes of making this a utility that has appeal and utility for broad (but still targeted) use cases, it feels worth it to make sure we're thinking about how we want this to be used before committing to one approach or another.

**Review by bucko13 — APPROVED** [review]
@waldenraines there are some other comments that are getting addressed in #41 so this one should be good to go.

**Pull request merged**
